### PR TITLE
use exponential backoff retry policy

### PR DESCRIPTION
**Pull request by d2r**
Use com.netflix.curator.retry.ExponentialBackoffRetry instead of com.netflix.curator.retry.RetryNTimes for ZK connections.

### CODE REVIEW

**Comment by nathanmarz** [comment]
Did you run into something in particular that required this?

**Comment by d2r** [comment]
Reiterating Andy's rationale for the change:

With a large cluster, after Zookeeper is not accessible for a period of time, we found Zookeeper can get overwhelmed when trying to recover.  The exponential back-off helps in those situations.

File: src/jvm/backtype/storm/utils/Utils.java [hunk 1]
```diff
@@ -299,7 +299,7 @@ public static CuratorFramework newCurator(Map conf, List<String> servers, Object
                     .connectString(zkStr)
                     .connectionTimeoutMs(Utils.getInt(conf.get(Config.STORM_ZOOKEEPER_CONNECTION_TIMEOUT)))
                     .sessionTimeoutMs(Utils.getInt(conf.get(Config.STORM_ZOOKEEPER_SESSION_TIMEOUT)))
-                    .retryPolicy(new RetryNTimes(Utils.getInt(conf.get(Config.STORM_ZOOKEEPER_RETRY_TIMES)), Utils.getInt(conf.get(Config.STORM_ZOOKEEPER_RETRY_INTERVAL))));
+                    .retryPolicy(new ExponentialBackoffRetry(Utils.getInt(conf.get(Config.STORM_ZOOKEEPER_RETRY_TIMES)), Utils.getInt(conf.get(Config.STORM_ZOOKEEPER_RETRY_INTERVAL))));
```
Comment (nathanmarz) [inline src/jvm/backtype/storm/utils/Utils.java]: According to http://netflix.github.com/curator/doc/com/netflix/curator/retry/ExponentialBackoffRetry.html, these arguments are reversed here.

**Comment by nathanmarz** [comment]
This is a good idea in principle. Some questions:
1. What do the time intervals look like exactly? If your base is 100ms, what will the next 10 tries look like? 200ms, 400ms, 800ms, 1.6s, 3.2s, etc?
2. Is there a way to set a ceiling on the retry time? Is this even desirable? Seems like some serious problems could arise if a worker is waiting 30s to retry an operation – like for example, a worker timeout.

**Comment by d2r** [comment]
1. They would look like: 100~200ms, 100~400ms, 100~800ms, etc.  With a base of 100ms, the interval for a given `retryCount` is random between 100ms ~ 2^(`retryCount` + 1).  So it is randomized, with the upper limit increasing exponentially.
2. Yes we can, and I agree that probably would be a good idea. It defaults to `Integer.MAX_VALUE`.  What would be a good ceiling?

**Comment by d2r** [comment]
Further notes on the ceiling.  The version of curator we pull in now (1.0.1) does not have a ceiling.  We'd need at least 1.2.6.

Also, in the curator git revision comments there is a note that they changed the ceiling from 60000ms to `Integer.MAX_VALUE` to avoid "breaking existing code."

**Comment by d2r** [comment]
What we really want is BoundedExponentialBackoffRetry, which was added in curator 1.2.5.

But I note that there is a significant hurdle to pulling in a newer curator: #225, especially it's dependency on zookeeper 3.4.5 #467

**Comment by d2r** [comment]
Put a configurable ceiling on the backoff retry, defaults to 30s, but we can discuss what it should be.
Added JUnit and a test for the change.

Trying to avoid pulling in a new Curator.  We may not want to go down this road either...

**Comment by nathanmarz** [comment]
Yea, we've had some trouble with the new ZK client library that we're trying to hash out. 

As for the diff – let's keep the tests in Clojure. Just add a new util_test.clj file into test/clj/backtype/storm/

**Comment by d2r** [comment]
Sure, I can rewrite the test. 

Curious, what is the barrier to using JUnit tests?

**Comment by nathanmarz** [comment]
JUnit adds no additional functionality compared to what we can already do with clojure.test, so it's simpler to stick with the existing unit testing framework. This also reduces dependencies of the project and makes it simpler to run the unit tests (just lein test).

**Comment by d2r** [comment]
OK, rewritten with a clojure test.

File: test/clj/backtype/storm/util_test.clj [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        servers ["bogus_server"]
+        arbitrary_port 42
+        curator (Utils/newCurator conf servers arbitrary_port)
+        retry (-> curator .getZookeeperClient .getRetryPolicy)
+       ]
+    (is (.isAssignableFrom ExponentialBackoffRetry (.getClass retry)))
+    (is (= (.getBaseSleepTimeMs retry) expected_interval)) 
+    (is (= (.getN retry) expected_retries)) 
+    (is (= (.getMaxRetryInterval retry) expected_ceiling)) 
+    ; It would be very unlikely for this to fail three times.
```
Comment (nathanmarz) [inline test/clj/backtype/storm/util_test.clj]: What does this mean exactly? Isn't the function stateless, so can't you just test the value once?

**Comment by nathanmarz** [comment]
OK – I think we're close to finishing this. First off, let's get rid of the project.clj changes as those are no longer necessary. I also made one more comment in the code,

File: test/clj/backtype/storm/util_test.clj (hunk 2 above)
Comment (d2r) [inline test/clj/backtype/storm/util_test.clj]: The result is psuedo-randomly distributed, so there is a slim chance with these values that working code could trip the assertion.

I could trivially set the ceiling to be less than the base, and that would guarantee the assertion is true.

**Comment by d2r** [comment]
Reverted the project.clj change, sorry about that.  Replied to your in-line comment.

**Comment by nathanmarz** [comment]
Merged. I changed the config to explicitly say millis.

**Pull request merged**